R/projection_rules.R

```diff
@@ -0,0 +1,275 @@
+#' @param age_from age_from
+#' @param age_to age_to
+non_linear_scale_up <- function(d, year_from, year_to, endpoint, gender = NULL, age_from = NULL, age_to = NULL){
+  years <- seq(year_from, year_to, 1)
+  cov <- IA2030_projection(year_from-1, d$coverage[d$year == year_from -1], year_to, endpoint)
+  t <- data.frame(year = years, coverage = cov)
+  if(!is.null(gender)){
+    t$gender <- gender
+  }
+  if(!is.null(age_from)){
+    t$age_from <- age_from
+  }
+  if(!is.null(age_to)){
+    t$age_to <- age_to
+  }
+  dat <- dplyr::bind_rows(d, t)
+  return(dat)
+}
+
+#' campaign coverage projection rule - future follow-up campaigns
+#' this function is designed under WHO M/MR guidance
+#' depending on the most recent campaign, routine introduction date and routine coverage level of vaccine_base
+#' @param d a data frame containing at least (year, coverage)
+#' @param dat data frame containing at least vaccine_base upto year_to
+#' @param vaccine_base vaccine to be dependent on
+#' @param year_current year current
+#' @param year_to year to
+#' @param look_back evaluate follow-up sia frequency from year_current - look_back
+#' @param sia_level campaign coverage level
+#' @param age_from campaign target age from
+#' @param age_to campaign target age to
+#' @param  gender 1 for both, 2 for males, 3 for females
+sia_follow_up <- function(d, dat, vaccine_base, year_current, year_to, look_back = 4, sia_level = 0.9, age_from = 1, age_to = 5, gender = 1){
+  ## vaccine_base is a baseline vaccine for evaluating follow-up campaign frequency
+  ## evaluate vaccine_base levels from a baseline year
+  ## baseline year is determined by year_intro, year_current, and last_sia_year
+  ## if no historical sia, baseline_year = year_current or future_year_intro if applicable
+  ## if last_sia_year < year_current-look_back, baseline_year = year_current or future_year_intro if applicable
+  ## if last_sia_year > year_current-look_back & year_intro <= last_sia_year, baseline_year = last_sia_year
+  ## if last_sia_year > year_current-look_back & year_intro > last_sia_year, baseline_year = year_intro
+
+  ## otherwise baseline_year = year_current
+
+  ## MR follow-up sia frequency
+  ## every 2 years if mcv1 level is < 60%
+  ## every 3 years if mcv1 level is 60 - 80%
+  ## every 4 years if mcv1 level is >= 80% & < 95%
+  ## N.B. if projected sia didn't happen, i.e. not recorded as a historical sia, postpone to year_current+1
+
+  ## determine baseline year
+  b <- dat[dat$vaccine == vaccine_base, ]
+  y_int <- min(b$year)
+  y_current <- year_current
+
+  if(nrow(d) == 0L){
+    y_base <- max(y_int, y_current)
+  } else {
+    y_last_sia <- max(d$year)
+    if(y_last_sia < y_current - look_back){
+      y_base <- max(y_int, y_current)
+    } else if(y_last_sia > y_current - look_back & y_last_sia >= y_int){
+      y_base <- y_last_sia
+    } else {
+      y_base <- y_int
+    }
+  }
+
+  ## project future campaigns from y_base to year_to
+  i <- y_base
+  t <- NULL
```

Comment:
```suggestion
  year <- NULL
```

(and then below when used) so that later you go


```
  t <- data.frame(year = year,
                  coverage = sia_level,
                  age_from = age_from,
                  age_to = age_to,
                  gender = gender)
```

which will be easier to work with in a year's time

Reply:
Done.